### ISSUE TITLE
Strange bug regarding WithRemoteAuthentication

### ISSUE THREAD

**Issue by AndreSteenbergen** [Finbuckle/Finbuckle.MultiTenant#76]
I am using a social login provider (google in this case) All works well with the following configuration
```
            services.AddMultiTenant()
                .WithStrategy<TenantStrategy>()
                .AddPerTenantCookieAuthentication()
                //.WithRemoteAuthentication()
                .AddPerTenantSocialLogins()
                ;

            services.Replace(ServiceDescriptor.Singleton<IAuthenticationSchemeProvider, MultiTenantAuthenticationSchemeProvider>());
            //services.Replace(ServiceDescriptor.Scoped<IAuthenticationService, MultiTenantAuthenticationService>());
            services.TryAddSingleton<IRemoteAuthenticationStrategy, RemoteAuthenticationStrategy>();
```

WithRemoteAuthentication has 3 methods it calls. For some reason the replacement for `IAuthenticationService` makes the 'signin-google' route redirect to the refererrer instead of the callbackurl.

BTW, it works for me as well without the .WithRemote, I guess is it because I am using a custom host strategy.

WRONG:
![afbeelding](https://user-images.githubusercontent.com/7676790/46739722-7b958e80-cca1-11e8-813e-a5745fdb6d30.png)

CORRECT (with middle call not made)
![afbeelding](https://user-images.githubusercontent.com/7676790/46739952-f2328c00-cca1-11e8-824c-7c13b86bebbd.png)

The challenge seems ok, but then again, the challenge 302 is not human readable. My fix for me is to remove the WithRemoteAuthentication altogether, as it works without for me, and I don't understand what it's doing.

I hope there is enough info here, netcore 2.1

**Comment by AndrewTriesToCode** [comment]
Hi again, are you using Safari version 12 by any chance?

**Comment by AndrewTriesToCode** [comment]
Hi @AndreSteenbergen,
I'll try to see if I can explain the purpose of WithRemoteAuthentication (apologies if you already know some of this--I'm writing it in generally in case anyone else has the same question).

I'm not sure why it still works in your case without WithRemoteAuthentication--my guess is that it is setting a cookie, but that it isn't actually using the cookie-specific settings for the intended Tenant. Can you post an example project on a github repo so I can take a look? I tested the social logins using Facebook, but in theory it should work the same as Google.

**What is WithRemoteAuthentication doing?**
When you use a Social Login it's usually using OAuth2 internally. In OAuth2 you register "redirect urls" with the OAuth2 provider. This is the "/signin-google" route (or similar) that ASP.NET Core uses by default. When forwarding the user to the social provider you supply a redirect url which **usually** must match the registered redirect url. Sometimes for development purposes they will allow the submitted redirect url to not match the registered redirect url but this is **not** safe for production (I know Facebook allows this for dev). After the social sign in the user is redirected back to the redirect url that was passed in (which is usually the same one registered).

The problem is that this redirect url doesn't work with some of the multitenant strategies. When Google redirects you to the /signin-google page then we don't know which tenant it was intended for. You could register a separate redirect url for each tenant, but this would be cumbersome and OAuth2 does not require that implementers support more than one redirect url.

`WithRemoteAuthentication` registeres a custom `IAuthenticationService`, which handles this for you by encoding the tenant identifier into the OAuth2 state parameter, which is passed back to the redirect url after the user signs into Google. The custom `IAuthenticationService` also then decodes the original tenant identifier and sets the correct `TenantContext` so that the correct login cookie settings are used. For other operations the authentication service defers to the default ASP.NET Core behavior.

As an aside, I would warn against trying to use per tenant settings for the AppId and AppSecret on Social providers because this requires that the tenant register *your* redirect url on *their* developer account with the social login provider. In some cases you might have good reason to use the tenant's AppId and AppSecret on an OAuth2 provider, but for social logins it's almost always better to you *your* apps own AppId and AppSecret.

**A Safari 12 bug**
I'm not sure if you are using Safari 12, but it currently has an issue (maybe bug) where some OpenID Connect or OAuth2 flows don't work right because the same-site cookie policy is being applied aggressively on the redirect request to the redirection url (/signin-google). There is a workaround which I will try to post more info on later. More details on the bug [here](https://community.auth0.com/t/authentication-broken-on-asp-net-core-and-safari-on-ios-12-mojave/16077).

**Comment by AndreSteenbergen** [comment]
@achandlerwhite
I think my version works, because I am using host based multitenant. see also https://github.com/AndreSteenbergen/Aspnet.Identity.Akka
Maybe you can see what is going on in my project, why the extensionmethod does not work for me. The google login is configured and working for localhost:5000.

**Comment by AndrewTriesToCode** [comment]
@AndreSteenbergen 

Thanks Andre, your project looks interesting. I can see why it works for you--It looks like you have a default tenant selected if one is configured with a "*" identifier.

One other situation that `WithRemoteAuthentication` handles is if you want each tenant to have a different default authentication scheme. So if you want one tenant to default to google and another to facebook. By default this doesn't work because the ASP.NET Core `AuthenticationSchemeProvider` caches its scheme from the first call--so later a different tenant might get the wrong authentication scheme! If you skip `WithRemoteAuthentication` I recommend you test this scenario carefully.

Cheers!

**Comment by AndreSteenbergen** [comment]
The bug is more, logincallbacks and regular logins seem to not work when I use, `WithRemoteAuthentication` and set Tenantspecific cookies. So I stripped out that particular pieces of configuration. And it seems to work now, where it previously didn't.

Feel free to copy any code for a sample. I based this code on the externallogins example, I didn't like the large ConfigureServices method, so I created something simpler, and create some extensions methods to move it out of the startup.cs.

**Comment by AndrewTriesToCode** [comment]
Hi again @AndreSteenbergen, I still haven't fully figured this out.

I have a working example of using the Google social sign in for authentication (it will be merged into the next pull request and be part of a 2.0 release).

One difference between your code and my example though is that you are also using Identity. I have a working sample that also uses Identity but not one that uses Identity **and** the social logins together so I will try that next to see if I can reproduce the problem.

Also, in your code it is the "Profile" project where you are using multitenant, right? I want to make sure I'm looking at the right area.

**Comment by AndreSteenbergen** [comment]
@achandlerwhite To reproduce the issue you can simply fork my repo. It is not dependant on any database. It's simply pull and run (and stub the email provider). When you add the WithRemoteAuthentication and Cookie Per Tenant, you will see the loginCallbacks won't be called, and the CallbackUrl's are correctly specified. The google login works for localhost:5000, all others aren't configured.

**Comment by AndrewTriesToCode** [comment]
When Google redirects the user to localhost:5000/signin-google the host string will always be "localhost". In your custom resolver it will always resolve to the default tenant. The result is that cookie will be set as if it is your default tenant, "Calqo BV", This means even if the user trying to authenticate is under the "Claxe" tenant the authentication will always treat it like it is under the "Calqo BV" tenant. This is what the `WithRemoteAuthentication` is supposed to fix.

Now, I think I do see a problem with my library--the returnUrl ASP.NET Core uses internally typically doesn't include the original host of the request so the redirect is not sent to the correct tenant. I will look into this further .

**Comment by AndreSteenbergen** [comment]
I don't understand the explanation. During testing I never leave localhost, so it should always resolve into the CalqoBV tenant. The callback redirect never ends up in the callback url at the moment. When using subdomains or completely different domains it doesn't matter that much, as cookies are domain specific. The extensions methods are doing something, I just can't think what might go wrong. Only thing I can think of Aspnet doesn't like the cookie name change. But that would mean a bug in aspnet identity.

**Comment by AndrewTriesToCode** [comment]
Hi @AndreSteenbergen ,
Good news, I think I found the problem. I apologize for all the back and forth and that some of my ideas were not in the right direction. Thanks for reporting the issue and providing all the details to me.

The problem is that my authentication service was not respecting pre-existing authentication properties sent to a a challenge. Previously I was doing
```
// properties is a method parameter
properties = new AuthenticationProperties();
properties.Items.Add("tenantIdentifier", multiTenantContext.TenantInfo.Identifier);
```
but now I have set it to
```
// properties is a method parameter
properties = properties ?? new AuthenticationProperties();
properties.Items.Add("tenantIdentifier", multiTenantContext.TenantInfo.Identifier);
```

Turns out ASP.Net Core Identity was passing in a some properties in the `properties` and I was erasing them.

This fix will be in the 2.0 release coming soon.
